File: OrcScala/src/orc/lib/comp/ComparisonSite.java
//
// ComparisonSite.java -- Java class ComparisonSite
// Project OrcScala
//
// $Id$
//
// Copyright (c) 2009 The University of Texas at Austin. All rights reserved.
//
// Use and redistribution of this file is governed by the license terms in
// the LICENSE file found in the project's top-level directory and also found at
// URL: http://orc.csres.utexas.edu/license.shtml .
//

package orc.lib.comp;

import java.math.BigDecimal;
import java.math.BigInteger;

import orc.error.runtime.ArgumentTypeMismatchException;
import orc.error.runtime.TokenException;
import orc.types.Type;
import orc.values.sites.TypedSite;
import orc.values.sites.compatibility.Args;
import orc.values.sites.compatibility.Args.NumericBinaryOperator;
import orc.values.sites.compatibility.EvalSite;
import orc.values.sites.compatibility.Types;

/**
 * @author quark
 */
@SuppressWarnings("synthetic-access")
public abstract class ComparisonSite extends EvalSite implements TypedSite {
	private static class MyOperator implements NumericBinaryOperator<Integer> {
		@Override
		public Integer apply(final BigInteger a, final BigInteger b) {
			return Integer.valueOf(a.compareTo(b));
		}

		@Override
		public Integer apply(final BigDecimal a, final BigDecimal b) {
			return Integer.valueOf(a.compareTo(b));
		}

		@Override
		public Integer apply(final int a, final int b) {
			return Integer.valueOf(a - b);
		}

		@Override
		public Integer apply(final long a, final long b) {
			return Integer.valueOf((int) (a - b));
		}

		@Override
		public Integer apply(final byte a, final byte b) {
			return Integer.valueOf(a - b);
		}

		@Override
		public Integer apply(final short a, final short b) {
			return Integer.valueOf(a - b);
		}

		@Override
		public Integer apply(final double a, final double b) {
			return Integer.valueOf(Double.compare(a, b));
		}

		@Override
		public Integer apply(final float a, final float b) {
			return Integer.valueOf(Float.compare(a, b));
		}
	}

	@Override
	public Object evaluate(final Args args) throws TokenException {
		final Object arg0 = args.getArg(0);
		final Object arg1 = args.getArg(1);
		if (arg0 instanceof Number && arg1 instanceof Number) {
			final int a = Args.applyNumericOperator((Number) arg0, (Number) arg1, new MyOperator()).intValue();
			return Boolean.valueOf(compare(a));
		} else {
			if (!(arg0 instanceof Comparable)) {
				throw new ArgumentTypeMismatchException(0, "Comparable<Object>", args.getArg(0).getClass().getCanonicalName());
			}
			@SuppressWarnings("unchecked")
			final int a = ((Comparable<Object>) arg0).compareTo(arg1);
			return Boolean.valueOf(compare(a));
		}
	}

	abstract public boolean compare(int a);

	@Override
	public Type orcType() {
		return Types.function(Types.top(), Types.top(), Types.bool());
	}
}

File: OrcScala/src/orc/lib/comp/ComparisonSite.scala
//
// ComparisonSite.scala -- Scala class ComparisonSite and objects Greater, Greq, Leq, Less, and Inequal
// Project OrcScala
//
// Created by amp on Sept 25, 2017.
//
// Copyright (c) 2018 The University of Texas at Austin. All rights reserved.
//
// Use and redistribution of this file is governed by the license terms in
// the LICENSE file found in the project's top-level directory and also found at
// URL: http://orc.csres.utexas.edu/license.shtml .
//

package orc.lib.comp

import orc.types.{ BooleanType, SimpleFunctionType, Top }
import orc.values.sites.{ FunctionalSite, OverloadedDirectInvokerMethod2, TalkativeSite, IllegalArgumentInvoker }

abstract class ComparisonSite extends OverloadedDirectInvokerMethod2[AnyRef, AnyRef] with FunctionalSite {
  def getInvokerSpecialized(arg1: AnyRef, arg2: AnyRef) = {
    // TODO: This does not handle all possible reasonable cases and some of the priorities are weird. When we improve the numeric stack we should fix this.
    (arg1, arg2) match {
      case (a: BigDecimal, b: BigDecimal) =>
        invoker(a, b)((a, b) => compare(a.compareTo(b)))
      case (a: java.lang.Double, b: java.lang.Double) =>
        invoker(a, b)((a, b) => compare(a.compareTo(b)))
      case (a: java.math.BigDecimal, b: java.math.BigDecimal) =>
        invoker(a, b)((a, b) => compare(a.compareTo(b)))
      case (a: java.lang.Long, b: java.lang.Long) =>
        invoker(a, b)((a, b) => compare(a.compareTo(b)))
      case (a: java.lang.Integer, b: java.lang.Integer) =>
        invoker(a, b)((a, b) => compare(a.compareTo(b)))
      case (a: BigInt, b: BigInt) =>
        invoker(a, b)((a, b) => compare(a.compareTo(b)))
      case (a: java.math.BigInteger, b: java.math.BigInteger) =>
        invoker(a, b)((a, b) => compare(a.compareTo(b)))
      case (a: BigDecimal, b: Number) =>
        invoker(a, b)((a, b) => compare(a.compareTo(b.doubleValue())))
      case (a: Number, b: BigDecimal) =>
        invoker(a, b)((a, b) => compare(-b.compareTo(a.doubleValue())))
      case (a: BigInt, b: Number) =>
        invoker(a, b)((a, b) => compare(a.compareTo(b.longValue())))
      case (a: Number, b: BigInt) =>
        invoker(a, b)((a, b) => compare(-b.compareTo(a.longValue())))
      case (a: java.lang.Double, b: Number) =>
        invoker(a, b)((a, b) => compare(a.compareTo(b.doubleValue())))
      case (a: Number, b: java.lang.Double) =>
        invoker(a, b)((a, b) => compare(a.doubleValue().compareTo(b)))
      case (a: Comparable[_], b: Comparable[_]) =>
        invoker(a, b)((a, b) =>
          try {
            compare(a.asInstanceOf[Comparable[AnyRef]].compareTo(b))
          } catch {
            case e: ClassCastException =>
              throw new IllegalArgumentException(s"$this($a, $b)", e)
          }
        )
      case (a: AnyRef, b: AnyRef) =>
        IllegalArgumentInvoker(this, Array(a, b))
    }
  }

  protected def compare(comp: Int): Boolean

  def orcType() = SimpleFunctionType(Top, Top, BooleanType)
}

case object Greater extends ComparisonSite with LocalSingletonSite {
  override def getInvokerSpecialized(arg1: AnyRef, arg2: AnyRef) = {
    // TODO: This does not handle all possible reasonable cases and some of the priorities are weird. When we improve the numeric stack we should fix this.
    (arg1, arg2) match {
      case (a: BigDecimal, b: BigDecimal) =>
        invoker(a, b)((a, b) => a > b)
      case (a: java.lang.Double, b: java.lang.Double) =>
        invoker(a, b)((a, b) => a > b)
      case (a: java.lang.Long, b: java.lang.Long) =>
        invoker(a, b)((a, b) => a > b)
      case (a: java.lang.Integer, b: java.lang.Integer) =>
        invoker(a, b)((a, b) => a > b)
      case (a: BigInt, b: BigInt) =>
        invoker(a, b)((a, b) => a > b)
      case _ =>
        super.getInvokerSpecialized(arg1, arg2)
    }
  }

  protected def compare(comp: Int): Boolean = comp > 0
  override def toString = "Greater"
}

case object Greq extends ComparisonSite with LocalSingletonSite {
  override def getInvokerSpecialized(arg1: AnyRef, arg2: AnyRef) = {
    // TODO: This does not handle all possible reasonable cases and some of the priorities are weird. When we improve the numeric stack we should fix this.
    (arg1, arg2) match {
      case (a: BigDecimal, b: BigDecimal) =>
        invoker(a, b)((a, b) => a >= b)
      case (a: java.lang.Double, b: java.lang.Double) =>
        invoker(a, b)((a, b) => a >= b)
      case (a: java.lang.Long, b: java.lang.Long) =>
        invoker(a, b)((a, b) => a >= b)
      case (a: java.lang.Integer, b: java.lang.Integer) =>
        invoker(a, b)((a, b) => a >= b)
      case (a: BigInt, b: BigInt) =>
        invoker(a, b)((a, b) => a >= b)
      case _ =>
        super.getInvokerSpecialized(arg1, arg2)
    }
  }

  protected def compare(comp: Int): Boolean = comp >= 0
  override def toString = "Greq"
}

case object Leq extends ComparisonSite with LocalSingletonSite {
  override def getInvokerSpecialized(arg1: AnyRef, arg2: AnyRef) = {
    // TODO: This does not handle all possible reasonable cases and some of the priorities are weird. When we improve the numeric stack we should fix this.
    (arg1, arg2) match {
      case (a: BigDecimal, b: BigDecimal) =>
        invoker(a, b)((a, b) => a <= b)
      case (a: java.lang.Double, b: java.lang.Double) =>
        invoker(a, b)((a, b) => a <= b)
      case (a: java.lang.Long, b: java.lang.Long) =>
        invoker(a, b)((a, b) => a <= b)
      case (a: java.lang.Integer, b: java.lang.Integer) =>
        invoker(a, b)((a, b) => a <= b)
      case (a: BigInt, b: BigInt) =>
        invoker(a, b)((a, b) => a <= b)
      case _ =>
        super.getInvokerSpecialized(arg1, arg2)
    }
  }

  protected def compare(comp: Int): Boolean = comp <= 0
  override def toString = "Leq"
}

case object Less extends ComparisonSite with LocalSingletonSite {
  override def getInvokerSpecialized(arg1: AnyRef, arg2: AnyRef) = {
    // TODO: This does not handle all possible reasonable cases and some of the priorities are weird. When we improve the numeric stack we should fix this.
    (arg1, arg2) match {
      case (a: BigDecimal, b: BigDecimal) =>
        invoker(a, b)((a, b) => a < b)
      case (a: java.lang.Double, b: java.lang.Double) =>
        invoker(a, b)((a, b) => a < b)
      case (a: java.lang.Long, b: java.lang.Long) =>
        invoker(a, b)((a, b) => a < b)
      case (a: java.lang.Integer, b: java.lang.Integer) =>
        invoker(a, b)((a, b) => a < b)
      case (a: BigInt, b: BigInt) =>
        invoker(a, b)((a, b) => a < b)
      case _ =>
        super.getInvokerSpecialized(arg1, arg2)
    }
  }

  protected def compare(comp: Int): Boolean = comp < 0
  override def toString = "Less"
}

case object Inequal extends OverloadedDirectInvokerMethod2[Any, Any] with FunctionalSite with TalkativeSite with LocalSingletonSite {
  override def name = "Inequal"

  def getInvokerSpecialized(a: Any, b: Any) = {
    invokerStaticType(a, b)((a, b) => {
      if (a == null)
        b != null
      else
        a != b
    })
  }

  def orcType() = SimpleFunctionType(Top, Top, BooleanType)
}

